Extensions/retags.js

```diff
@@ -19,111 +19,168 @@ XKit.extensions.retags = {
+			var nonPopoverNotes = $baseMutatedElements.find('.note:not(.ui_post_badge)');
+			if ($baseMutatedElements.hasClass('note-list')) {
+				//remove existing retags items (if tabbing between popover views)
+				$baseMutatedElements.find('.retags-wrapper').remove();
+				var elements = $baseMutatedElements.find('.post-activity-note');
+				notesProcessed = true;
+				XKit.extensions.retags.tag_popover(elements.filter(XKit.extensions.retags.selectors));
+			} else if ($baseMutatedElements.hasClass('post-activity-note') && !notesProcessed) {
+				XKit.extensions.retags.tag_popover($baseMutatedElements.filter(XKit.extensions.retags.selectors));
+			} else if (nonPopoverNotes.length > 0) {
+				XKit.extensions.retags.tag(nonPopoverNotes.filter(XKit.extensions.retags.selectors));
 			} else {
-				XKit.extensions.retags.tag($el.filter(XKit.extensions.retags.selectors));
+				XKit.extensions.retags.tag($baseMutatedElements.filter(XKit.extensions.retags.selectors));
 			}
 		});
 	}),
 
-	add_toggle: function(){
-		var toggle = 'toggle_' + this.blog_name;
-		if (XKit.browser().mobile) {
-			this.html_toggle.appendTo('.primary-nav');
-			XKit.tools.add_css('label.retags .binary_switch_label {font-size:15px; color:white; padding-bottom:15px; }','retags_mobile_label');
-		} else {
-			this.html_toggle.appendTo('.ui_notes_switcher .part-toggle');
-		}
-		$('#retags-toggle').change(function(){
-			if ($(this).prop('checked')) {
-				XKit.storage.set('retags', toggle, 'true');
-				XKit.extensions.retags.css_toggle.appendTo('head');
-				if (XKit.browser().mobile) {
-					XKit.extensions.retags.mobile_toggle.appendTo('head');
+	tag_popover: function(elements){
+		$(elements).each(function(){
+			var $element = $(this), retagClass, $target, url;
+			if ($element.find('div.retags').length) {
+				return false;
+			}
+			if ($element.hasClass('is_reply') || $element.hasClass('is_answer') || $element.hasClass('type_6')) {
+				return;
+			}
+			// popover
+			if ($element.hasClass('note-item')) {
+				retagClass = 'with_commentary';
+				$target = $element;
+				var peepr_data = $target.find('.note-text-link').data('peepr');
+				url = '';
+				if (peepr_data) {
+					url = 'http://' + peepr_data.tumblelog.trim() + '.tumblr.com/post/' + peepr_data.postId.trim();
+				}
+			} 
+			if (url) {
+				url = url.split('/');
+				var host = url[2], id = url[4],	cache = 'post_' + id;
+				if (XKit.storage.get('retags', cache, null) !== null) {
+					var retags = XKit.extensions.retags.build_tag_collection($element, retagClass, XKit.storage.get('retags', cache));
+					if (!retags) {
+						return;
+					}
+					retags = "<li class='retags-wrapper'>" + retags[0].outerHTML + "</li>";
+					$target.after(retags);
+				} else {
+					var requestUrl = 'https://api.tumblr.com/v2/blog/' + host + '/posts/info?id=' + id + '&api_key=' + XKit.extensions.retags.api_key;
+					XKit.extensions.retags.request($element, retagClass, cache, requestUrl)
+						.then(function(retags) {
```

Comment:
The cache should only be used at one level of abstraction—either entirely inside this function or entirely outside of it. right now we mix the two, leading to a leaky abstraction and weird side-effects. I would recommend encapsulating the caching logic inside this function.